app/drivers/kscan/kscan_gpio_matrix.c

```diff
@@ -1,48 +1,155 @@
 /*
- * Copyright (c) 2020 The ZMK Contributors
+ * Copyright (c) 2020-2021 The ZMK Contributors
  *
  * SPDX-License-Identifier: MIT
  */
 
-#define DT_DRV_COMPAT zmk_kscan_gpio_matrix
-
 #include <device.h>
-#include <drivers/kscan.h>
+#include <devicetree.h>
 #include <drivers/gpio.h>
+#include <drivers/kscan.h>
 #include <logging/log.h>
+#include <sys/__assert.h>
+#include <sys/util.h>
 
 LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
 
+#define DT_DRV_COMPAT zmk_kscan_gpio_matrix
+
 #if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
 
-struct kscan_gpio_item_config {
-    char *label;
+#define INST_DIODE_DIR(n) DT_ENUM_IDX(DT_DRV_INST(n), diode_direction)
+#define COND_DIODE_DIR(n, row2col_code, col2row_code)                                              \
+    COND_CODE_0(INST_DIODE_DIR(n), row2col_code, col2row_code)
+
+#define INST_ROWS_LEN(n) DT_INST_PROP_LEN(n, row_gpios)
+#define INST_COLS_LEN(n) DT_INST_PROP_LEN(n, col_gpios)
+#define INST_MATRIX_LEN(n) (INST_ROWS_LEN(n) * INST_COLS_LEN(n))
+#define INST_INPUTS_LEN(n) COND_DIODE_DIR(n, (INST_COLS_LEN(n)), (INST_ROWS_LEN(n)))
+
+#define USE_POLLING IS_ENABLED(CONFIG_ZMK_KSCAN_MATRIX_POLLING)
```

Comment:
Not needed *this PR*, but the reality, I'm not convinced this really should be a global Kconfig setting, versus a per-driver instance property. Will review more to see if that really *matters* in the scope of these changes or not.

Reply:
Yeah, I had considered that, but decided to just match the existing behavior for now.